synchronizer/synchronizer.go

```diff
@@ -599,23 +605,60 @@ func (s *ClientSynchronizer) checkTrustedState(batch state.Batch, tBatch *state.
 	return false
 }
 
-func (s *ClientSynchronizer) processSequenceBatches(sequencedBatches []etherman.SequencedBatch, blockNumber uint64, dbTx pgx.Tx) error {
+func (s *ClientSynchronizer) processSequenceBatches(sequencedBatches []etherman.SequencedBatch, dbTx pgx.Tx) error {
 	if len(sequencedBatches) == 0 {
 		log.Warn("Empty sequencedBatches array detected, ignoring...")
 		return nil
 	}
 
-	prevTimestamp, err := s.state.GetLastBatchTime(s.ctx, dbTx)
-	if err != nil {
-		log.Warn("Error fetching previous timestamp")
-		return err
-	}
 	for _, sbatch := range sequencedBatches {
+		// Ensure the L1 origin for this batch is in the database. Since the L1 origin assigned by
+		// HotShot is not necessarily the same as an L1 block which we added to the database as a
+		// result of receiving a contract event, it might not be.
+		blockNumber := sbatch.BlockNumber
+		exists, err := s.state.ContainsBlock(s.ctx, blockNumber, dbTx)
+		if err != nil {
+			log.Errorf("error fetching L1 block %d from db: %v", blockNumber, err)
+			rollbackErr := dbTx.Rollback(s.ctx)
+			if rollbackErr != nil {
+				log.Fatalf("error rolling back state. BlockNumber: %d, rollbackErr: %s, error: %w", blockNumber, rollbackErr.Error(), err)
+			}
+			return err
+		}
+		if !exists {
+			header, err := s.etherMan.HeaderByNumber(s.ctx, big.NewInt(int64(blockNumber)))
+			if err != nil {
+				log.Errorf("error fetching L1 block %d: %v", blockNumber, err)
+				rollbackErr := dbTx.Rollback(s.ctx)
+				if rollbackErr != nil {
+					log.Fatalf("error rolling back state. BlockNumber: %d, rollbackErr: %s, error: %w", blockNumber, rollbackErr.Error(), err)
+				}
+				return err
+			}
+			b := state.Block{
+				BlockNumber: blockNumber,
+				BlockHash:   header.Hash(),
+				ParentHash:  header.ParentHash,
+				ReceivedAt:  time.Unix(int64(header.Time), 0),
+			}
+			log.Infof("L1 block %d does not already exist, adding to database", blockNumber)
+			err = s.state.AddBlock(s.ctx, &b, dbTx)
+			if err != nil {
```

Comment:
What is the behavior if the contract event handler is concurrently calling AddBlock? Is that a race condition we need to address here? If the contract handler adds the block between lines 619 and here and 645 errors out, we wouldn't need to bubble up an error.

Reply:
This should not be possible because we hold a lock on the DB here (everything done with `dbTx` should be atomic).

Also, `AddBlock` is idempotent, so even if this race did happen, 645 should succeed and everything should be ok.